First step for adding OOP to Mozci. Adding Status and related classes.

vaibhavmagarwal opened the pull request.

vaibhavmagarwal commented:
As discussed on IRC, the OOP model to be followed - https://etherpad.mozilla.org/mozci-ooo
r? @armenzg

armenzg commented:
I'm going to look at this now.

armenzg commented:
My apologies in advance as I might at times change what I have said before.
It's only now that I see the code that I can better picture things.

Feel free to disagree with any of what I say.

armenzg commented on mozci/mozci.py:
```diff
@@ -64,14 +64,15 @@ def _status_summary(jobs):
-        status = buildapi.query_job_status(job)
-        if status == buildapi.PENDING:
+        status_obj = buildapi.BuildapiStatus()
+        status = status_obj.query_job_status(job)
```
Could you please call it instead BuildapiJob? Otherwise, it reads as the status of buildapi, the self-serve service.
Could you make BuildapiJob have a constructor that receives a job dict?
That class should have a "status" property.
If "status" is not set, we call query_job_status() (lazy loading).

Could you also make BuildapiJob not to inherit from Status()?
Also instead of Status() we should name it StatusCodes() to make it sound like we're dealing with predefined codes. It feels weird to use an obj to store constants but it will make it easier for when we have TC support.

I'm also OK to leave the status codes as normal constants for now since I would kick myself if TC support will be substantially different and these changes would be unnecessary.

That way you can write something like this:
status_codes = buildapi.StatusCodes()
for job in jobs:
   job = buildapi.BuildapiJob(job)
   if job.status == status_codes.pending # Note that I don't use upper case

vaibhavmagarwal replied:
Hi Armen,

I would like to differ in some points mentioned above. I think class BuildapiJobStatus() makes the most sense since we are really querying for the status of the job and not the whole job itself. In future, when we query the status from TH, I will name it as TreeherderJobStatus().  Also, I think "status" should be a property of Job class (which I will make in next PR) and not of BuildapiJobStatus. We are querying BuildapiJobStatus only once in this function for *each* job, thus query_job_status() will be queried every time, and lazy loading will really not take place in this function. I agree to other points, and am right now keeping status codes as normal constants. These are my thoughts, and I am open for more discussion.

armenzg commented:
This code landing affects this PR:
https://github.com/armenzg/mozilla_ci_tools/pull/250

Would you prefer to keep the buildapi status codes as constants until we see what happens with TC?

vaibhavmagarwal commented:
r? @armenzg 
Armen, I have updated the code and kept in mind the lazy loading for status that we may need in future. Therefore, I have introduced a get_status() method in the class. Please let me know if this makes sense. Thanks!

armenzg commented:
Oh my goodness! It looks much nicer than I thought it could look.
